### ISSUE TITLE
Virtual Function

### ISSUE THREAD

**Issue by Udayv56** [Udayv56/CPP#39]
#include<iostream>
using namespace std;
 
class BaseClass
{
  public:
    int var1;
    virtual void display(){
        cout<<var1<<endl;
        cout<<"BASE CLASS"<<endl;
    }
};

class DerivedClass : public BaseClass{
    public:
    int var2;
    void display(){
        cout<<var1<<" "<<var2<<endl;
        cout<<"DERIVED CLASS"<<endl;
        
    }
};

int main(){
    BaseClass* base_class_pointer;
    BaseClass obj_base;
    DerivedClass obj_derived;
    base_class_pointer = &obj_derived; 
    base_class_pointer->display(); // default behaviour baseclass 
    // to make it run derived one then use virtual 
    BaseClass* base_class_pointer2;
    base_class_pointer2 = &obj_base;
    base_class_pointer2->display();    
    return 0;
}

**OUTPUT-**
0 0
DERIVED CLASS
0
BASE CLASS

In the given code, we have two classes: `BaseClass` and `DerivedClass`. `DerivedClass` is derived from `BaseClass`, which means it inherits all the members (variables and functions) of the base class. 

The `BaseClass` has a member variable `var1` and a virtual function `display()`. The `display()` function simply outputs the value of `var1` and prints "BASE CLASS" on the console.

The `DerivedClass` also has a member variable `var2`, and it overrides the `display()` function of the base class. The overridden `display()` function in the derived class outputs the values of both `var1` and `var2` and prints "DERIVED CLASS" on the console.

In the `main()` function, we declare a pointer `base_class_pointer` of type `BaseClass*`, an object `obj_base` of type `BaseClass`, and an object `obj_derived` of type `DerivedClass`.

Then, we assign the address of `obj_derived` to the `base_class_pointer` using the statement `base_class_pointer = &obj_derived;`. This is possible because the derived class object can be assigned to a pointer of its base class. 

Next, we call the `display()` function using the pointer `base_class_pointer`. Since the `display()` function is marked as virtual in the base class, it enables dynamic dispatch or late binding. It means that the function call is resolved at runtime based on the type of the object being pointed to. In this case, the actual object being pointed to is of type `DerivedClass`, so the overridden `display()` function in the `DerivedClass` is called, and it prints the values of `var1` and `var2` along with "DERIVED CLASS" on the console.

To demonstrate the default behavior of the base class, we create another pointer `base_class_pointer2` of type `BaseClass*` and assign the address of `obj_base` to it using the statement `base_class_pointer2 = &obj_base;`. Here, the actual object being pointed to is of type `BaseClass`, so when we call the `display()` function using `base_class_pointer2`, it calls the base class `display()` function and outputs the value of `var1` along with "BASE CLASS".

In summary, virtual functions in C++ allow dynamic dispatch or late binding, meaning the appropriate function to call is determined at runtime based on the type of the object being pointed to. By marking a function as virtual in the base class, you can ensure that the correct function is called even when using a pointer of the base class to refer to objects of derived classes. This is useful in achieving polymorphic behavior and enabling function overriding in derived classes.
